docs/developer/developing-plugins.md

```diff
@@ -0,0 +1,148 @@
+
+Here is the pattern typically followed by most application and transport plugins for receiving and validating a configuration:
+
+```go
+package main
+
+
+[...]
+
+type configT struct {
+    Address string `yaml:"address" validate:"required"` //required parameter
+    Instance string `yaml:"instance" validate:"oneof=inst0 inst1 inst2 inst3"` //must be element in set
+    Port int `yaml:"port"`
+}
+
+type Socket struct {
+    conf configT
+    logger *logging.Logger
+}
+
+func (s *Socket) Config(c []byte) error {
+    s.conf = configT{}
+    err := config.ParseConfig(bytes.NewReader(c), &s.conf)
+    if err != nil {
+        return err // sg-core will log configuration error message and exit
+    }
+    return nil
+}
+
+[...]
+
+func New(l *logging.Logger) transport.Transport {
+    return &Socket{
+        logger: l,
+        conf: configT{
+            Port: 9090 // set defaults. Only overriden if input data for Config() contains the port option
+        }
+    }
+}
+
+```
+
+## Transports
+
+Transport plugins listen on an external messaging protocal and deliver received messages to handlers that have been bound to it per the administrator's configuration. They receive a configuration block from the main configuration file and write to handlers by involing transport.WriteFn in Run().
+
+Transports should contain the minimal amount of code necessary to fulfill this functionality. 
+
+Transport plugin objects must implement the the Transport interface:
+```go
+type Transport interface {
+	Config([]byte) error
+	Run(context.Context, *sync.WaitGroup, transport.WriteFn, chan bool)
+	Listen(data.Event)
+}
+```
+
+## Handlers
+
+Handlers parse incoming blobs from the transport into objects and delivers those objects to the internal buses. There are two types of handlers: metric handlers and event handlers. Metric handlers deliver metric objects to the internal metrics bus while event handlers deliver event objects to the internal events bus. These metrics and events are then consumed by the application plugins.
+
+Handlers must remain simple: they take in no configuration and should only handle one message type. See the [collectd-metrics](https://github.com/pleimer/sg-core-refactor/tree/master/plugins/handler/collectd-metrics) plugin for an example. Additionally, handlers should not print logs (this is why no logger is passed into the New() function). If errors occur while parsing messages, handlers should create their own metrics or events recording the error(s) and write them to the bus. The `collectd-metrics` handler iterates a counter every time a parsing error occurs and submits the number as a metric to the metric bus.
```

Comment:
Kind of agree with Chris here. I don't see advantage of generating logs in handler itself too. There will have to be logic in each application processing the specific handler messages. If author of any application will forget to handle errors (or simply will not care), the operators might loose important information about failures. I'm not saying that handlers should not send error messages, just that failures should IMO be logged also on handler layer.

Reply:
I see the concern, it definitely leaves information out. The reason I chose to suggest against it is that because handlers are dealing with a large quantity of messages, the logs could become extremely unwieldy very quickly. If the developer could be responsible enough to log the first error, and all proceeding duplicates of the error ignore, maybe it is worth it. What do you guys think?

@paramite In most cases I don't think the Application should contain logic to deal with specific metrics but rather should only contain the logic to deal with the [metric type](https://github.com/pleimer/sg-core/blob/482f18b8301cea535a950e20a91ebda507bdf4a6/pkg/data/data.go#L28). For events, maybe so if the plugin needs to react to them. In the case of parser errors in the collectd-metrics handler, the handler creates a new metric with a counter specifying the number of parsing errors that have occurred [here](https://github.com/pleimer/sg-core/blob/482f18b8301cea535a950e20a91ebda507bdf4a6/plugins/handler/collectd-metrics/main.go#L53). The Prometheus plugin doesn't do anything special with this metric and instead just stores it like all the rest. But, we can plot this metric in Grafana and see if any parsing errors are occurring without filling up the logs.

Reply:
Maybe the handler could return an error in this case and the manager could then unbind it from the transport. This may be the better option and wouldn't leave so much up to the plugin developer